### PR TITLE
ENH: add distance along the Hilbert Curve

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
This is a numpy-based implementation of the distance along the Hilbert Curve, similar as we have in [dask-geopandas](https://github.com/geopandas/dask-geopandas/blob/main/dask_geopandas/hilbert_distance.py) but thus not using numba but a vectorized implementation based on the version in GEOS (which is basically https://github.com/rawrunprotected/hilbert_curves)

With the version of GEOS / https://github.com/rawrunprotected/hilbert_curves, it was actually trivially to have it run vectorized on a numpy array of discrete coordinates as well (literally the same code).

I am directly creating a PR in geopandas here, and not in dask-geopandas, since this is a feature we were thinking to move to base geopandas anyway (we can use it here as well to sort geometries).

This still needs tests, docs, and hooking it up in the public API, etc.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Based on a first test, this new implementation gives the same result as the dask-geopandas one, while actually being faster. On 1 million points, I get ca 500ms for the numba-based implementation of dask-geopandas vs ~80ms for this numpy-based implementation. 
(this is only for the lower-level encoding of discrete coordinates to distance, so the actual difference starting from a GeoSeries will be smaller)

Notebook: https://notebooksharing.space/view/fb83908941339cc7654dbe271c02b2df49eaf8cc4d434077da58066f9cd1ff86#displayOptions=

**Comment by martinfleis** [comment]
I suggest using this implementation in `dask-geopandas` as well (before we do 0.1 release) as it allows us dropping `numba` as a dependency.

**Comment by jorisvandenbossche** [comment]
> I suggest using this implementation in dask-geopandas as well (before we do 0.1 release) as it allows us dropping numba as a dependency.

Ah, yes, that's a good point. I think we can duplicate it on the short term, and later remove it again from dask-geopandas to rely on the geopandas version.

**Comment by brendan-ward** [comment]
What about adding this in Shapely instead of GeoPandas?  This seems more tightly coupled to working with geometries than to GeoPandas specifically, and may be easier to leverage anything from GEOS C API if this becomes available there and is more performant than the numpy solution.

File: geopandas/tools/hilbert_curve.py
Comment (brendan-ward) [inline geopandas/tools/hilbert_curve.py:7]:
```suggestion
    Calculate distance along a Hilbert curve a GeoDataFrame
```

File: geopandas/tools/hilbert_curve.py
Comment (brendan-ward) [inline geopandas/tools/hilbert_curve.py:8]:
```suggestion
```
(unclear on what `int coordinates` is supposed to refer to)

File: geopandas/tools/hilbert_curve.py
Comment (brendan-ward) [inline geopandas/tools/hilbert_curve.py:16]: Elsewhere (GEOS, [Wikipedia](https://en.wikipedia.org/wiki/Hilbert_curve)) refers to this as `level` instead.  Suggest using `level` here.

File: geopandas/tools/hilbert_curve.py
Comment (brendan-ward) [inline geopandas/tools/hilbert_curve.py:20]:
```suggestion
    Pandas Series containing Hilbert distances
```

**Comment by martinfleis** [comment]
@brendan-ward I think that long-term, that is the best solution. I think that the main benefit of having it here right now is that we don't have to wait for shapely to release it and then expose it here about 3 releases from now.

File: geopandas/tools/hilbert_curve.py
Comment (brendan-ward) [inline geopandas/tools/hilbert_curve.py:64]:
```suggestion
    # Transform continuous value to discrete integer for each dimension
```

(midpoints are floating point)

**Review by brendan-ward — COMMENTED** [review]
A few minor docstring comments, but otherwise looks good.

(side note: the codecov messages make it quite a bit harder to review code, maybe those should be turned off or handled differently during review)

File: geopandas/tools/hilbert_curve.py
Comment (martinfleis) [reply geopandas/tools/hilbert_curve.py:16]: We also wanted to rename `p` in dask-geopandas. We thought about `precision` there but `level` could be fine as well. I guess that works for all options (Hilbert, Morton, Geohash?).

**Comment by brendan-ward** [comment]
Simplification and likely speedup: instead of calculating the `total_bounds` of the input geometries, calculate the extent (in numpy) of their midpoints calculated here.  This should also lead to better distribution of geometries along the curve for geometries that have large extents (since it would be scaling their midpoints relative to each other rather than midpoints to the total bounds of the original geometries).

**Comment by jorisvandenbossche** [comment]
> (side note: the codecov messages make it quite a bit harder to review code, maybe those should be turned off or handled differently during review)

Yeah, those are indeed annoying (it's because I didn't add any tests ..). Small hint that you can easily hide them all at once with:

Screenshot from 2022-01-28 09-03-43

**Comment by jorisvandenbossche** [comment]
Thanks for the review @brendan-ward !

So I am first going to update https://github.com/geopandas/dask-geopandas/pull/134 with those comments, have that merged, and then just copy it over here (as mentioned by Martin, short term I think it's good to include this code in those several places, to make it easier to start using it, agreed that long term it would fit in shapely).

> Simplification and likely speedup: instead of calculating the `total_bounds` of the input geometries, calculate the extent (in numpy) of their midpoints calculated here. This should also lead to better distribution of geometries along the curve for geometries that have large extents (since it would be scaling their midpoints relative to each other rather than midpoints to the total bounds of the original geometries).

Yes, I actually already had a commit locally (pushed it now) that calculate the "total bounds" from the bounds, which we access anyway (and the underlying implementation of `total_bounds` in pygeos/shapely is exactly that, getting the `bounds` and calculating the min/max of that, so we were effectively extracting the bounds twice). 
But indeed I can take that a step further and only do this "total bounds" calculation after we calculated the midpoints.

File: geopandas/tools/hilbert_curve.py
Comment (jorisvandenbossche) [reply geopandas/tools/hilbert_curve.py:16]: Yes, for me `level` is fine as well.

File: geopandas/tools/hilbert_curve.py
Comment (brendan-ward) [inline geopandas/tools/hilbert_curve.py:66]: Note that in [GEOS this is calculated](https://github.com/libgeos/geos/blob/main/src/shape/fractal/HilbertEncoder.cpp#L32) as `(2 ** p)-1`

Comment (brendan-ward) [reply geopandas/tools/hilbert_curve.py:66]: > Did you notice that when comparing the results you got with GEOS?

Yes, that's how I spotted this.

In the [GEOS test](https://github.com/libgeos/geos/pull/556/commits/72caa8d833901db161a6471ec20e9a4b735bfd49#diff-7067e7dc083a60e06851744edf1d72c9e92b47585c1a941c719eea87f14cb790R61) I was using `POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))` at a level of 2, and in GeoPandas it yields 8 and in GEOS yields 2; when I use `(2 ** p)-1` in GeoPandas they match.

Comment (jorisvandenbossche) [reply geopandas/tools/hilbert_curve.py:66]: I corrected this in https://github.com/geopandas/dask-geopandas/pull/134 (will update here once it is finalized / merged in dask-geopandas)

**Comment by jorisvandenbossche** [comment]
I think this was actually ready.

**Comment by martinfleis** [comment]
Can you add a changelog note? Otherwise this seems to be ready.

**Review by martinfleis — APPROVED** [review]
Thanks!

**Pull request merged**
